### PR TITLE
BUG: Fixed Von Mises distribution for big values of kappa

### PR DESCRIPTION
**Pull request by bashtage**
Extends branch in #17378 to prevent fix applying to RandomState and changing its stream.

### CODE REVIEW

File: numpy/random/src/distributions/distributions.c [hunk 1]
```diff
@@ -853,9 +854,30 @@ double random_vonmises(bitgen_t *bitgen_state, double mu, double kappa) {
...
+        result = mu + sqrt(1. / kappa) * random_standard_normal(bitgen_state);
+        /* Check if result is within bounds */
+        if (result < -M_PI) {
+          return result + 2*M_PI;
+        }
+        if (result > M_PI) {
+          return result - 2*M_PI;
+        }
+        return result;
+      } else {
```
Comment (charris) [inline numpy/random/src/distributions/distributions.c:872]: Indentation needs to be 4 spaces, `else` on new line.

File: numpy/random/src/distributions/distributions.c [hunk 2]
```diff
@@ -853,9 +854,30 @@ double random_vonmises(bitgen_t *bitgen_state, double mu, double kappa) {
        */
       s = (1. / kappa + kappa);
     } else {
-      double r = 1 + sqrt(1 + 4 * kappa * kappa);
-      double rho = (r - sqrt(2 * r)) / (2 * kappa);
-      s = (1 + rho * rho) / (2 * rho);
+#ifndef NP_RANDOM_LEGACY
+      /* Fallback to normal distribution for big values of kappa
```
Comment (charris) [inline numpy/random/src/distributions/distributions.c:858]: First comment line should be blank.

File: numpy/random/src/distributions/distributions.c (hunk 1 above)
Comment (charris) [reply numpy/random/src/distributions/distributions.c:872]: Style needs fixing all round, let's leave that to a later cleanup.

**Comment by rkern** [comment]
I think that it would be safer/more readable/more convenient to do the `#ifdef NP_RANDOM_LEGACY` switch between two copies of the whole function. Then we have free reign in the non-legacy copy of the function. Since we're not maintaining the legacy version, there's no DRY benefit to sharing the (thus-far) common parts of the function. I think it would make the task of verifying that the legacy variant remains unchanged much, much simpler. And it would be an easy convention that new contributors can follow, so we don't have to police them with followup PRs like this.

**Comment by charris** [comment]
The code coverage failure should probably be fixed as it seems a common case.

**Comment by bashtage** [comment]
> I think that it would be safer/more readable/more convenient to do the `#ifdef NP_RANDOM_LEGACY` switch between two copies of the whole function. Then we have free reign in the non-legacy copy of the function. Since we're not maintaining the legacy version, there's no DRY benefit to sharing the (thus-far) common parts of the function. I think it would make the task of verifying that the legacy variant remains unchanged much, much simpler. And it would be an easy convention that new contributors can follow, so we don't have to police them with followup PRs like this.

I moved it to legacy-distributions.c as legacy_vonmises.  This has the advantage of clarifying that these functions should not be altered except under special circumstances.

File: numpy/random/tests/test_generator_mt19937.py [hunk 3]
```diff
@@ -1735,6 +1735,26 @@ def test_vonmises_nan(self):
...
+            assert_allclose(random_state_vals, gen_vals)
+        else:
+            assert np.all(random_state_vals != gen_vals)
+
+    @pytest.mark.parametrize("mu", [-7., -np.pi, -3.1, np.pi, 3.2])
+    @pytest.mark.parametrize("kappa", [1e-9, 1e-6, 1, 1e3, 1e15])
+    def test_vonmises_large_kappa_range(self, mu, kappa):
+        r = random.vonmises(mu, kappa, 50)
+        assert_(np.all(r > -np.pi) and np.all(r <= np.pi))
+
```
Comment (bashtage) [inline numpy/random/tests/test_generator_mt19937.py:1757]: I think that coverage is a bit suspicious.  I added this test to ensure that all code paths are hit.

File: numpy/random/src/distributions/distributions.c (hunk 1 above)
Comment (bashtage) [reply numpy/random/src/distributions/distributions.c]: While I don't mind changing the commit, the entire file is formatted using these conventions (2 spaces, single like else).  IIRC think this is Google C format as formatted using clang-format.

**Comment by rkern** [comment]
> I moved it to legacy-distributions.c as legacy_vonmises.

Agreed.

**Comment by charris** [comment]
The azure failure is unrelated, I didn't rerun the test as the output would be lost.

**Comment by charris** [comment]
Thanks @bashtage .

**Pull request merged**